cmd/nerdctl/build_test.go

```diff
@@ -48,6 +48,104 @@ CMD ["echo", "nerdctl-build-test-string"]
 	base.Cmd("run", "--rm", imageName).AssertOutExactly("nerdctl-build-test-string\n")
 }
 
+func TestBuildBaseImage(t *testing.T) {
+	if os.Getenv("TEST_BUILDKIT_WORKER") != "containerd" {
+		t.Skip()
+	}
+	t.Parallel()
+	testutil.RequiresBuild(t)
+	base := testutil.NewBase(t)
+	imageName := testutil.Identifier(t)
+	defer base.Cmd("rmi", imageName).Run()
+	imageName2 := imageName + "-2"
+	defer base.Cmd("rmi", imageName2).Run()
+
+	dockerfile := fmt.Sprintf(`FROM %s
+RUN echo hello > /hello
+CMD ["echo", "nerdctl-build-test-string"]
+	`, testutil.CommonImage)
+
+	buildCtx, err := createBuildContext(dockerfile)
+	assert.NilError(t, err)
+	defer os.RemoveAll(buildCtx)
+
+	base.Cmd("build", "-t", imageName, buildCtx).AssertOK()
+	base.Cmd("build", buildCtx, "-t", imageName).AssertOK()
+
+	dockerfile2 := fmt.Sprintf(`FROM %s
+RUN echo hello2 > /hello2
+CMD ["cat", "/hello2"]
+	`, imageName)
+
+	buildCtx2, err := createBuildContext(dockerfile2)
+	assert.NilError(t, err)
+	defer os.RemoveAll(buildCtx2)
+
+	base.Cmd("build", "-t", imageName2, buildCtx2).AssertOK()
+	base.Cmd("build", buildCtx2, "-t", imageName2).AssertOK()
+
+	base.Cmd("run", "--rm", imageName2).AssertOutExactly("hello2\n")
+}
+
+func TestBuildFromContainerd(t *testing.T) {
+	if os.Getenv("TEST_BUILDKIT_WORKER") != "containerd" {
+		t.Skip()
+	}
+	t.Parallel()
+	testutil.RequiresBuild(t)
+	base := testutil.NewBase(t)
+	imageName := testutil.Identifier(t)
+	defer base.Cmd("rmi", imageName).Run()
+	imageName2 := imageName + "-2"
+	defer base.Cmd("rmi", imageName2).Run()
+
+	// FIXME: BuildKit sometimes tries to use base image manifests of platforms that hasn't been
+	//        pulled by `nerdctl pull`. This leads to "not found" error for the base image.
+	//        To avoid this issue, images shared to BuildKit should always be pulled by manifest
+	//        digest or `--all-platforms` needs to be added.
+	base.Cmd("pull", "--all-platforms", testutil.CommonImage).AssertOK()
+	base.Cmd("tag", testutil.CommonImage, imageName).AssertOK()
+	base.Cmd("rmi", testutil.CommonImage).AssertOK()
+
+	dockerfile2 := fmt.Sprintf(`FROM %s
+RUN echo hello2 > /hello2
+CMD ["cat", "/hello2"]
+	`, imageName)
+
+	buildCtx2, err := createBuildContext(dockerfile2)
+	assert.NilError(t, err)
+	defer os.RemoveAll(buildCtx2)
+
+	base.Cmd("build", "-t", imageName2, buildCtx2).AssertOK()
+	base.Cmd("build", buildCtx2, "-t", imageName2).AssertOK()
+
+	base.Cmd("run", "--rm", imageName2).AssertOutExactly("hello2\n")
+}
+
+func TestBuildUnlazy(t *testing.T) {
```

Comment:
What is "Unlazy" here? Doesn't seem related to stargz stuff though

Reply:
Deleted this test because TestBuild already coverts this.

What I wanted to test was the following edge case:

If a build doesn't run any commands nor perform COPY on the base image, BuildKit tries not to download that image (i.e. keeps it lazy). This behaviour happens not only with stargz-snapshotter but also with any type of snapshotter.
`type=oci`/`type=docker` triggers unlazying of that base image but `type=image` doesn't. This leads to "not found" error for the contents of that image e.g. when nerdctl tries to run it. As a workaround, nerdctl adds `unpack=true` to `buildctl build` which triggers unlazy even when `type=image` is specified.

TestBuild is exactly the situation where the base image can be lazy. So this test is enough for testing that this workaround actually works.